File: Java/src/_0043_Multiply_Strings/Solution.java
package _0043_Multiply_Strings;

/**
 * Date: 2020/8/30 9:23
 * Content:
 */
class Solution {

    // 两个字符串相加
    private String add(String num1, String num2) {

        int p = num1.length() - 1;
        int q = num2.length() - 1;

        StringBuilder res = new StringBuilder();
        int temp = 0;
        while (p >= 0 || q >= 0) {
            int x = (p >= 0 ? (num1.charAt(p) - '0') : 0);
            int y = (q >= 0 ? (num2.charAt(q) - '0') : 0);
            int t = x + y + temp;
            res.append((t) % 10);
            temp = (t) / 10;
            p--;
            q--;
        }
        if (temp != 0) res.append(temp);

        return res.reverse().toString();
    }

    // 一个字符串和一个数字字符相乘
    public String multiply(String num1, char ch) {
        if (ch == '0') return "0";
        else if (ch == '1') return num1;
        else {
            int x = ch - '0';
            int p = num1.length() - 1;

            StringBuilder res = new StringBuilder();
            int temp = 0;
            while (p >= 0) {
                int y = num1.charAt(p) - '0';
                int t = x * y + temp;
                res.append((t) % 10);
                temp = (t) / 10;
                p--;
            }
            if (temp != 0) res.append(temp);
            return res.reverse().toString();
        }
    }

    // 两个字符串相乘
    public String multiply(String num1, String num2) {

        String smallNum = (num1.length() < num2.length() ? num1 : num2);
        String largeNum = (num1.length() < num2.length() ? num2 : num1);
        StringBuilder suffix = new StringBuilder();
        String res = "0";
        for (int i = smallNum.length() - 1; i >= 0; i--) {
            String mul = multiply(largeNum, smallNum.charAt(i)) + suffix;
            res = add(mul, res);
            suffix.append('0');
        }
        return res;
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).multiply("123", "456"));  // 56088
    }
}

File: Python/_0043_Multiply_Strings/Solution.py
# coding=utf-8
# Date: 2021/5/22 11:11


# 执行用时：88 ms, 在所有 Python3 提交中击败了65.28%的用户
# 内存消耗：15.1 MB, 在所有 Python3 提交中击败了8.38%的用户
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        n = len(num1); m = len(num2)
        A = []; B = []
        for i in range(n - 1, -1, -1):
            A.append(ord(num1[i]) - ord('0'))
        for i in range(m - 1, -1, -1):
            B.append(ord(num2[i]) - ord('0'))
        C = [0 for _ in range(n + m)]
        for i in range(n):
            for j in range(m):
                C[i + j] += A[i] * B[j]
        t = 0
        for i in range(len(C)):
            t += C[i]
            C[i] = t % 10
            t //= 10

        # 处理输出
        k = len(C) - 1
        while k > 0 and C[k] == 0:
            k -= 1
        res = ""
        while k >= 0:
            res += str(C[k])
            k -= 1
        return res


if __name__ == "__main__":
    print(Solution().multiply("2", "3"))  # 6
    print(Solution().multiply("123", "456"))  # 56088
